File: backend/app/schemas/attendance.py
from typing import Optional

from pydantic import BaseModel, Field

from app.schemas._common import IdStr


class CheckInRequest(BaseModel):
    conferenceUid: IdStr


class AttendanceOut(BaseModel):
    status: str
    markedOn: Optional[str] = None
    distanceMeters: Optional[float] = None


class VerifyLocationRequest(BaseModel):
    conferenceUid: IdStr
    latitude: float = Field(ge=-90, le=90)
    longitude: float = Field(ge=-180, le=180)


class VerifyLocationOut(BaseModel):
    distanceMeters: Optional[float] = None
    withinRadius: Optional[bool] = None
    # The geofence radius this check was made against, in metres (defaults to
    # 100). Lets the trainee app tell the user exactly how close they need to be.
    radiusMeters: Optional[int] = None
    venueLabel: Optional[str] = None

File: src/api/attendance.ts
import { apiRequest, apiUpload } from "./client";

export type AttendanceRecord = {
  status: string;
  markedOn: string | null;
  distanceMeters: number | null;
};

export function checkIn(token: string, conferenceUid: string) {
  return apiRequest<AttendanceRecord>("/attendance/check-in", {
    method: "POST",
    headers: { Authorization: `Bearer ${token}` },
    body: JSON.stringify({ conferenceUid }),
  });
}

export type VerifyLocationResult = {
  distanceMeters: number | null;
  withinRadius: boolean | null;
  /** Geofence radius this check was made against, in metres (defaults to 100). */
  radiusMeters: number | null;
  venueLabel: string | null;
};

export function verifyLocation(token: string, conferenceUid: string, latitude: number, longitude: number) {
  return apiRequest<VerifyLocationResult>("/attendance/verify-location", {
    method: "POST",
    headers: { Authorization: `Bearer ${token}` },
    body: JSON.stringify({ conferenceUid, latitude, longitude }),
  });
}

export type SecureCheckInPayload = {
  conferenceUid: string;
  latitude: number;
  longitude: number;
  photo: { uri: string; name: string; type: string };
};

export function secureCheckIn(token: string, payload: SecureCheckInPayload) {
  const formData = new FormData();
  formData.append("conferenceUid", payload.conferenceUid);
  formData.append("latitude", String(payload.latitude));
  formData.append("longitude", String(payload.longitude));
  formData.append("photo", {
    uri: payload.photo.uri,
    name: payload.photo.name,
    type: payload.photo.type,
  } as unknown as Blob);

  return apiUpload<AttendanceRecord>("/attendance/check-in/secure", formData, token);
}

export { ApiError } from "./client";
